### PR TITLE
[clang-tidy] Add bsl-type-decl-header

### PR DESCRIPTION
**Pull request by ytuo**
A2-9-1
A header file name shall be identical to a type name declared in it if it declares a type.

### CODE REVIEW

File: clang-tools-extra/clang-tidy/bsl/TypeDeclHeaderCheck.cpp [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang {
+namespace tidy {
+namespace bsl {
+
+void TypeDeclHeaderCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(tagDecl().bind("x"), this);
```
Comment (cjams) [inline clang-tools-extra/clang-tidy/bsl/TypeDeclHeaderCheck.cpp]: Should this be namedDecl instead of tagDecl?

Comment (ytuo) [reply clang-tools-extra/clang-tidy/bsl/TypeDeclHeaderCheck.cpp]: I'm trying to match classes, structs, enums, and unions with this. I'll need to add typedefs though

File: clang-tools-extra/test/clang-tidy/checkers/BSLTypeDeclHeaderTypedef.hpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
+// RUN: %check_clang_tidy %s bsl-type-decl-header %t
+
+typedef struct {
+  int a;
+  int b;
+} BSLTypeDeclHeaderTypedef; // Compliant
+
+typedef long NotBSLTypeDeclHeaderTypedef; // Non-compliant
+// CHECK-MESSAGES: [[@LINE-1]]:7: warning:
```
Comment (cjams) [inline clang-tools-extra/test/clang-tidy/checkers/BSLTypeDeclHeaderTypedef.hpp:9]: @ytuo is this missing the text from the warning?

**Comment by rianquinn** [comment]
This has been added via a different check

**Pull request closed without merging**